ops/model.py

```diff
@@ -603,17 +603,30 @@ def open_port(self, protocol: typing.Literal['tcp', 'udp', 'icmp'],
         `Juju documentation <https://juju.is/docs/sdk/hook-tool#heading--open-port>`__
         for more detail.
 
+        Use :meth:`set_ports` for a more declarative approach where all of
+        the ports that should be open are provided in a single call.
+
         Args:
             protocol: String representing the protocol; must be one of
                 'tcp', 'udp', or 'icmp' (lowercase is recommended, but
                 uppercase is also supported).
             port: The port to open. Required for TCP and UDP; not allowed
                 for ICMP.
+
+        Raises:
+            ModelError: If ``port`` is provided when ``protocol`` is 'icmp'
+                or ``port`` is not provided when ``protocol`` is 'tcp' or
+                'udp'.
         """
-        self._backend.open_port(protocol.lower(), port)
+        normalised_protocol : str = protocol.lower()
+        if normalised_protocol == 'icmp' and port is not None:
+            raise ModelError("icmp cannot have a port number specified")
+        elif normalised_protocol in ('tcp', 'udp') and port is None:
+            raise ModelError(f"{normalised_protocol} must have a port number specified")
+        self._backend.open_port(normalised_protocol, port)
```

Comment:
Sorry if I'd muddied the waters earlier, but I think we should drop this code change (but we can keep the "Raises: ModelError" docstring change). This checking is automatically provided by the hook tool, which will return an error and `ModelError` will be raised, so I don't think we should duplicate those checks here.

The `testing._TestingModelBackend.open_port` function also does this check, but that's to simulate the hook tool.

Same comment for `close_port`.